File: src/test/java/kitchenpos/application/fake/FakeMenuRepository.java
package kitchenpos.application.fake;

import kitchenpos.domain.Menu;
import kitchenpos.domain.MenuRepository;

import java.util.*;
import java.util.stream.Collectors;

public class FakeMenuRepository implements MenuRepository {
    Map<UUID, Menu> memory = new HashMap<>();
    @Override
    public List<Menu> findAllByIdIn(List<UUID> ids) {
        return memory.values().stream()
                .anyMatch(menu -> ids.contains(menu.getId())) ? new ArrayList<>(memory.values()) : Collections.emptyList();
    }

    @Override
    public Optional<Menu> findById(UUID id) {
        return Optional.ofNullable(memory.get(id));
    }

    @Override
    public Menu save(Menu menu) {
        memory.put(menu.getId(), menu);
        return memory.get(menu.getId());
    }

    @Override
    public List<Menu> findAll() {
        return new ArrayList<>(memory.values());
    }

    @Override
    public List<Menu> findAllByProductId(UUID id) {
        List<Menu> result = new ArrayList<>();
        memory.values()
                .forEach(menu -> menu.getMenuProducts()
                        .stream()
                        .anyMatch(product -> product.getProductId().equals(id) ? result.add(menu) : false));
        return result;
    }
}

File: src/test/kotlin/kitchenpos/application/fake/FakeMenuRepository.kt
package kitchenpos.application.fake

import kitchenpos.domain.Menu
import kitchenpos.domain.MenuRepository
import java.util.*

class FakeMenuRepository : MenuRepository {
    private val menus = mutableMapOf<UUID, Menu>()

    override fun save(entity: Menu): Menu {
        menus[entity.id] = entity
        return entity
    }

    override fun findById(id: UUID): Optional<Menu> {
        return Optional.ofNullable(menus[id])
    }

    override fun findAllByIdIn(ids: MutableList<UUID>): MutableList<Menu> {
        return menus.filter {
            it.key in ids
        }.map {
            it.value
        }.toMutableList()
    }

    override fun findAllByProductId(productId: UUID?): MutableList<Menu> {
        return menus.filter { menu ->
            menu.value.menuProducts.map {
                    menuProduct ->
                menuProduct.productId
            }.contains(productId)
        }.map {
            it.value
        }.toMutableList()
    }

    override fun findAll(): MutableList<Menu> {
        return menus.values.toMutableList()
    }
}
